CHANGES.rst

```diff
@@ -300,6 +300,11 @@ astropy.extern
 astropy.io.ascii
 ^^^^^^^^^^^^^^^^
 
+  - Options not available with the fast reader raise a ``ParameterError``
+    if the latter was explicitly requested; ``guess=True`` will try
```

Comment:
I don't understand the part about `guess=True` will try all formats with a fast reader, nor the similar statement in the narrative docs.  In the trace below I made a file that will fail reading and thus show all the formats in order that were tried.  I did this with your branch and with 2.0.1 and the results are the same.  Maybe I missed it, but I don't see anything in the code diffs that would change the guess order (which is a good thing!).  The main real code changes I see are:
- Fix bug where fast_reader dict values are consumed during guessing
- Change API so that supplying `fast_reader` as a dict implies `'force'`.
- Lots of new testing (excellent!).
- Improved exception reporting
- Change `get_reader` so it returns a fast reader if `fast_reader` is supplied and evals to True in boolean context.
```
In [14]: cat junk.dat
a b
1 2
3 4 5

In [15]: ascii.read('junk.dat')
---------------------------------------------------------------------------
InconsistentTableError                    Traceback (most recent call last)
/Users/aldcroft/git/astropy/astropy/io/ascii/ui.py in _guess(table, read_kwargs, format, fast_reader)
    510             reader = get_reader(**read_kwargs)
--> 511             dat = reader.read(table)
    512             _read_trace.append({'kwargs': read_kwargs,

/Users/aldcroft/git/astropy/astropy/io/ascii/core.py in read(self, table)
   1193 
-> 1194                     raise InconsistentTableError(errmsg)
   1195 

InconsistentTableError: Number of header columns (2) inconsistent with data columns (3) at data line 1
Header values: ['a', 'b']
Data values: ['3', '4', '5']

During handling of the above exception, another exception occurred:

InconsistentTableError                    Traceback (most recent call last)
<ipython-input-15-7341372a4ff1> in <module>()
----> 1 ascii.read('junk.dat')

/Users/aldcroft/git/astropy/astropy/io/ascii/ui.py in read(table, guess, **kwargs)
    331         # through below to the non-guess way so that any problems result in a
    332         # more useful traceback.
--> 333         dat = _guess(table, new_kwargs, format, fast_reader_param)
    334         if dat is None:
    335             guess = False

/Users/aldcroft/git/astropy/astropy/io/ascii/ui.py in _guess(table, read_kwargs, format, fast_reader)
    539                    '************************************************************************']
    540             lines.extend(msg)
--> 541             raise core.InconsistentTableError('\n'.join(lines))
    542 
    543 

InconsistentTableError: 
ERROR: Unable to guess table format with the guesses listed below:
Reader:Ecsv fill_values: [('', '0')] strict_names: True
Reader:FixedWidthTwoLine fill_values: [('', '0')] strict_names: True
Reader:FastBasic fill_values: [('', '0')] strict_names: True
Reader:Basic fill_values: [('', '0')] strict_names: True
Reader:Rdb fill_values: [('', '0')] strict_names: True
Reader:FastTab fill_values: [('', '0')] strict_names: True
Reader:Tab fill_values: [('', '0')] strict_names: True
Reader:Cds fill_values: [('', '0')] strict_names: True
Reader:Daophot fill_values: [('', '0')] strict_names: True
Reader:SExtractor fill_values: [('', '0')] strict_names: True
Reader:Ipac fill_values: [('', '0')] strict_names: True
Reader:Latex fill_values: [('', '0')] strict_names: True
Reader:AASTex fill_values: [('', '0')] strict_names: True
Reader:FastCommentedHeader delimiter: '|' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastCommentedHeader delimiter: '|' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastCommentedHeader delimiter: ',' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastCommentedHeader delimiter: ',' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastCommentedHeader delimiter: ' ' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastCommentedHeader delimiter: ' ' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastCommentedHeader delimiter: '\\s' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastCommentedHeader delimiter: '\\s' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:CommentedHeader delimiter: '|' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:CommentedHeader delimiter: '|' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:CommentedHeader delimiter: ',' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:CommentedHeader delimiter: ',' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:CommentedHeader delimiter: ' ' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:CommentedHeader delimiter: ' ' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:CommentedHeader delimiter: '\\s' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:CommentedHeader delimiter: '\\s' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastBasic delimiter: '|' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastBasic delimiter: '|' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastBasic delimiter: ',' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastBasic delimiter: ',' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastBasic delimiter: ' ' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastBasic delimiter: ' ' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastBasic delimiter: '\\s' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastBasic delimiter: '\\s' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:Basic delimiter: '|' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:Basic delimiter: '|' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:Basic delimiter: ',' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:Basic delimiter: ',' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:Basic delimiter: ' ' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:Basic delimiter: ' ' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:Basic delimiter: '\\s' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:Basic delimiter: '\\s' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastNoHeader delimiter: '|' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastNoHeader delimiter: '|' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastNoHeader delimiter: ',' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastNoHeader delimiter: ',' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastNoHeader delimiter: ' ' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastNoHeader delimiter: ' ' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastNoHeader delimiter: '\\s' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastNoHeader delimiter: '\\s' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:NoHeader delimiter: '|' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:NoHeader delimiter: '|' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:NoHeader delimiter: ',' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:NoHeader delimiter: ',' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:NoHeader delimiter: ' ' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:NoHeader delimiter: ' ' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:NoHeader delimiter: '\\s' fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:NoHeader delimiter: '\\s' fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:Basic fill_values: [('', '0')]

************************************************************************
** ERROR: Unable to guess table format with the guesses listed above. **
**                                                                    **
** To figure out why the table did not read, use guess=False and      **
** appropriate arguments to read().  In particular specify the format **
** and any known attributes like the delimiter.                       **
************************************************************************
```

Reply:
The point was what guess order (or better perhaps sequence, as in the list of guesses included, as the order itself indeed does not and should not change) is adopted, if `fast_reader` is made mandatory - either explicitly via `'force'`, or implicitly by specifying a `fast_reader`. Instead trying to read your file with
```
>>> ascii.read('junk.dat', fast_reader={'exponent_style': 'a'})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/sw/lib/python2.7/site-packages/astropy/io/ascii/ui.py", line 333, in read
    dat = _guess(table, new_kwargs, format, fast_reader_param)
  File "/sw/lib/python2.7/site-packages/astropy/io/ascii/ui.py", line 541, in _guess
    raise core.InconsistentTableError('\n'.join(lines))
astropy.io.ascii.core.InconsistentTableError: 
ERROR: Unable to guess table format with the guesses listed below:
Reader:FastBasic fast_reader: {} fill_values: [('', '0')] strict_names: True
Reader:FastTab fast_reader: {} fill_values: [('', '0')] strict_names: True
Reader:FastCommentedHeader delimiter: '|' fast_reader: {} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastCommentedHeader delimiter: '|' fast_reader: {} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastCommentedHeader delimiter: ',' fast_reader: {} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastCommentedHeader delimiter: ',' fast_reader: {} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastCommentedHeader delimiter: ' ' fast_reader: {} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastCommentedHeader delimiter: ' ' fast_reader: {} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastCommentedHeader delimiter: '\\s' fast_reader: {'exponent_style': 'a'} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastCommentedHeader delimiter: '\\s' fast_reader: {'exponent_style': 'a'} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastBasic delimiter: '|' fast_reader: {} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastBasic delimiter: '|' fast_reader: {} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastBasic delimiter: ',' fast_reader: {} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastBasic delimiter: ',' fast_reader: {} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastBasic delimiter: ' ' fast_reader: {} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastBasic delimiter: ' ' fast_reader: {} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastBasic delimiter: '\\s' fast_reader: {'exponent_style': 'a'} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastBasic delimiter: '\\s' fast_reader: {'exponent_style': 'a'} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastNoHeader delimiter: '|' fast_reader: {} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastNoHeader delimiter: '|' fast_reader: {} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastNoHeader delimiter: ',' fast_reader: {} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastNoHeader delimiter: ',' fast_reader: {} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastNoHeader delimiter: ' ' fast_reader: {} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastNoHeader delimiter: ' ' fast_reader: {} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastNoHeader delimiter: '\\s' fast_reader: {'exponent_style': 'a'} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastNoHeader delimiter: '\\s' fast_reader: {'exponent_style': 'a'} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:Basic fast_reader: {} fill_values: [('', '0')]

************************************************************************
** ERROR: Unable to guess table format with the guesses listed above. **
**                                                                    **
** To figure out why the table did not read, use guess=False and      **
** appropriate arguments to read().  In particular specify the format **
** and any known attributes like the delimiter.                       **
************************************************************************
```
in this branch, whereas omitting only the change in [`ui.py:420`](https://github.com/astropy/astropy/pull/5578/files/f12dd5bb0aa272d4c8b2c136a21c9c41f483114b#diff-e6b2d1f1997a9b7e467819e2c98be409) lets the guesses start, as in your example, with
```
>>>
ERROR: Unable to guess table format with the guesses listed below:
Reader:Ecsv fast_reader: {} fill_values: [('', '0')] strict_names: True
Reader:FixedWidthTwoLine fast_reader: {} fill_values: [('', '0')] strict_names: True
Reader:FastBasic fast_reader: {} fill_values: [('', '0')] strict_names: True
...
```

Reply:
That said, this particular change might not be crucial for the actual results, i.e. in case of a file where the `fast_reader` _does_ return different results.

I. While the sequence of readers is different, the file is ultimately read in correctly, as long as the `fast_reader` dict is protected.

With '2.0.3':
```
>>> cat fort.dat                                                                                                                                       
A	B
1.D0	2.d0
1.D1	2.D1
>>> ascii.read('fort.dat')
<Table length=2>
 A    B  
str4 str4
---- ----
1.D0 2.d0
1.D1 2.D1
>>> ascii.get_read_trace()
[{'status': 'InconsistentTableError: ECSV header line like "# %ECSV <version>" not found as first line.  This is required for a ECSV file.', 'dt': '0.196 ms', 'kwargs': {'strict_names': True, 'fill_values': [('', '0')], 'Reader': <class 'astropy.io.ascii.ecsv.Ecsv'>}}, {'status': 'InconsistentTableError: Position line should only contain delimiters and one other character, e.g. "--- ------- ---".', 'dt': '0.191 ms', 'kwargs': {'strict_names': True, 'fill_values': [('', '0')], 'Reader': <class 'astropy.io.ascii.fixedwidth.FixedWidthTwoLine'>}}, {'status': 'ValueError: Strict name guessing requires at least two columns', 'dt': '0.112 ms', 'kwargs': {'strict_names': True, 'fill_values': [('', '0')], 'Reader': <class 'astropy.io.ascii.fastbasic.FastBasic'>}}, {'status': 'ValueError: Strict name guessing requires at least two columns', 'dt': '0.216 ms', 'kwargs': {'strict_names': True, 'fill_values': [('', '0')], 'Reader': <class 'astropy.io.ascii.basic.Basic'>}}, {'status': "ValueError: RDB types definitions do not all match [num](N|S): ['1.D0', '2.d0']", 'dt': '0.197 ms', 'kwargs': {'strict_names': True, 'fill_values': [('', '0')], 'Reader': <class 'astropy.io.ascii.basic.Rdb'>}}, {'status': 'Success (guessing)', 'dt': '1.248 ms', 'kwargs': {'strict_names': True, 'fill_values': [('', '0')], 'Reader': <class 'astropy.io.ascii.fastbasic.FastTab'>}}]
>>> ascii.read('fort.dat', fast_reader='force')
<Table length=2>
 A    B  
str4 str4
---- ----
1.D0 2.d0
1.D1 2.D1
>>> ascii.get_read_trace()
[{'status': 'ValueError: Strict name guessing requires at least two columns', 'dt': '0.131 ms', 'kwargs': {'fast_reader': 'force', 'strict_names': True, 'fill_values': [('', '0')], 'Reader': <class 'astropy.io.ascii.fastbasic.FastBasic'>}}, {'status': 'Success (guessing)', 'dt': '1.215 ms', 'kwargs': {'Reader': <class 'astropy.io.ascii.fastbasic.FastTab'>, 'strict_names': True, 'fill_values': [('', '0')], 'fast_reader': 'force'}}]
>>> ascii.read('fort.dat', fast_reader={'exponent_style': 'a'})
<Table length=2>
 A    B  
str4 str4
---- ----
1.D0 2.d0
1.D1 2.D1
>>> ascii.get_read_trace()
[{'status': 'InconsistentTableError: ECSV header line like "# %ECSV <version>" not found as first line.  This is required for a ECSV file.', 'dt': '0.283 ms', 'kwargs': {'Reader': <class 'astropy.io.ascii.ecsv.Ecsv'>, 'strict_names': True, 'fill_values': [('', '0')], 'fast_reader': {}}}, {'status': 'InconsistentTableError: Position line should only contain delimiters and one other character, e.g. "--- ------- ---".', 'dt': '0.206 ms', 'kwargs': {'fast_reader': {}, 'strict_names': True, 'fill_values': [('', '0')], 'Reader': <class 'astropy.io.ascii.fixedwidth.FixedWidthTwoLine'>}}, {'status': 'ValueError: Strict name guessing requires at least two columns', 'dt': '0.122 ms', 'kwargs': {'fast_reader': {}, 'strict_names': True, 'fill_values': [('', '0')], 'Reader': <class 'astropy.io.ascii.fastbasic.FastBasic'>}}, {'status': 'ValueError: Strict name guessing requires at least two columns', 'dt': '0.265 ms', 'kwargs': {'fast_reader': {}, 'strict_names': True, 'fill_values': [('', '0')], 'Reader': <class 'astropy.io.ascii.basic.Basic'>}}, {'status': "ValueError: RDB types definitions do not all match [num](N|S): ['1.D0', '2.d0']", 'dt': '0.207 ms', 'kwargs': {'Reader': <class 'astropy.io.ascii.basic.Rdb'>, 'strict_names': True, 'fill_values': [('', '0')], 'fast_reader': {}}}, {'status': 'Success (guessing)', 'dt': '1.187 ms', 'kwargs': {'Reader': <class 'astropy.io.ascii.fastbasic.FastTab'>, 'strict_names': True, 'fill_values': [('', '0')], 'fast_reader': {}}}]
```
With this PR, the last variant becomes
```
>>> ascii.read('fort.dat', fast_reader={'exponent_style': 'a'})
<Table length=2>
   A       B   
float64 float64
------- -------
    1.0     2.0
   10.0    20.0
>>> ascii.get_read_trace()
[{'status': 'ValueError: Strict name guessing requires at least two columns', 'dt': '0.137 ms', 'kwargs': {'fast_reader': {}, 'strict_names': True, 'fill_values': [('', '0')], 'Reader': <class 'astropy.io.ascii.fastbasic.FastBasic'>}}, {'status': 'Success (guessing)', 'dt': '1.205 ms', 'kwargs': {'Reader': <class 'astropy.io.ascii.fastbasic.FastTab'>, 'strict_names': True, 'fill_values': [('', '0')], 'fast_reader': {}}}]
```
Without the change in `ui.py:420` this becomes
```
>>> ascii.read('fort.dat', fast_reader={'exponent_style': 'a'})
<Table length=2>
   A       B   
float64 float64
------- -------
    1.0     2.0
   10.0    20.0
>>> ascii.get_read_trace()
[{'status': "ParameterError: fast_reader required with {'exponent_style': 'a'}, but this is not a fast C reader: <class 'astropy.io.ascii.ecsv.Ecsv'>", 'dt': '0.054 ms', 'kwargs': {'Reader': <class 'astropy.io.ascii.ecsv.Ecsv'>, 'strict_names': True, 'fill_values': [('', '0')], 'fast_reader': {}}}, {'status': "ParameterError: fast_reader required with {'exponent_style': 'a'}, but this is not a fast C reader: <class 'astropy.io.ascii.fixedwidth.FixedWidthTwoLine'>", 'dt': '0.030 ms', 'kwargs': {'fast_reader': {}, 'strict_names': True, 'fill_values': [('', '0')], 'Reader': <class 'astropy.io.ascii.fixedwidth.FixedWidthTwoLine'>}}, {'status': 'ValueError: Strict name guessing requires at least two columns', 'dt': '0.115 ms', 'kwargs': {'fast_reader': {}, 'strict_names': True, 'fill_values': [('', '0')], 'Reader': <class 'astropy.io.ascii.fastbasic.FastBasic'>}}, {'status': "ParameterError: fast_reader required with {'exponent_style': 'a'}, but this is not a fast C reader: <class 'astropy.io.ascii.basic.Basic'>", 'dt': '0.053 ms', 'kwargs': {'fast_reader': {'exponent_style': 'a'}, 'strict_names': True, 'fill_values': [('', '0')], 'Reader': <class 'astropy.io.ascii.basic.Basic'>}}, {'status': "ParameterError: fast_reader required with {'exponent_style': 'a'}, but this is not a fast C reader: <class 'astropy.io.ascii.basic.Rdb'>", 'dt': '0.043 ms', 'kwargs': {'Reader': <class 'astropy.io.ascii.basic.Rdb'>, 'strict_names': True, 'fill_values': [('', '0')], 'fast_reader': {'exponent_style': 'a'}}}, {'status': 'Success (guessing)', 'dt': '1.198 ms', 'kwargs': {'Reader': <class 'astropy.io.ascii.fastbasic.FastTab'>, 'strict_names': True, 'fill_values': [('', '0')], 'fast_reader': {}}}]
```
i.e. this saves a number of unneeded read attempts, but does not change the output. I guess I had a hypothetical case in mind where, say, `Reader:Rdb` would produce a valid result before the `fast_reader` got a chance to try, but such a case may not exist in reality.

II. With the change to `fast_reader` always being a `dict` in `3.x`, this particular issue seems already resolved.